tests/test_s3parallelcopy.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+import os
+import sys
+import unittest
+import uuid
+from concurrent.futures import ThreadPoolExecutor
+
+import boto3
+import botocore
+
+pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))  # noqa
+sys.path.insert(0, pkg_root)  # noqa
+
+from dss import stepfunctions
+from dss.stepfunctions import s3copyclient
+from dss.stepfunctions.s3copyclient.implementation import LAMBDA_PARALLELIZATION_FACTOR
+from dss.util.aws import AWS_MIN_CHUNK_SIZE
+from tests import eventually, infra
+from tests.infra import testmode
+
+
+class TestS3ParallelCopy(unittest.TestCase):
+    @staticmethod
+    def upload_part(bucket: str, key: str, upload_id: str, part_number: int):
+        s3_client = boto3.client("s3")
+        etag = s3_client.upload_part(
+            Bucket=bucket,
+            Key=key,
+            UploadId=upload_id,
+            PartNumber=part_number,
+            ContentLength=AWS_MIN_CHUNK_SIZE,
+            Body=chr(part_number % 256) * (AWS_MIN_CHUNK_SIZE))['ETag']
+        return etag
+
+    @eventually(30 * 60, 5.0, {AssertionError, botocore.exceptions.ClientError})
+    def _check_dst_key_etag(self, bucket: str, key: str, expected_etag: str):
+        s3_client = boto3.client("s3")
+        obj_metadata = s3_client.head_object(
+            Bucket=bucket,
+            Key=key
+        )
+        self.assertEqual(expected_etag, obj_metadata['ETag'].strip("\""))
+
+    @testmode.integration
+    def test_large_copy(self, num_parts=LAMBDA_PARALLELIZATION_FACTOR + 1):
+        test_bucket = infra.get_env("DSS_S3_BUCKET_TEST")
+        test_src_key = infra.generate_test_key()
+        s3_client = boto3.client("s3")
+        mpu = s3_client.create_multipart_upload(Bucket=test_bucket, Key=test_src_key)
+        part_jobs = list()
+
+        with ThreadPoolExecutor(max_workers=8) as tpe:
```

Comment:
I would use `8 * multiprocessing.cpu_count()` here. Since 90% of what these threads will be doing is wait for I/O, you could use significantly more threads than you have cores.

Reply:
In my experience running this, I saturated the network connection and more threads did not help.  I am more concerned with the memory footprint at this point.